cvat-ui/src/reducers/export-reducer.ts

```diff
@@ -0,0 +1,103 @@
+// Copyright (C) 2021 Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+
+import { ExportActions, ExportActionTypes } from 'actions/export-actions';
+import getCore from 'cvat-core-wrapper';
+import { ExportState } from './interfaces';
+
+const core = getCore();
+
+const defaultState: ExportState = {
+    tasks: {
+        datasets: {},
+        annotations: {},
+    },
+    projects: {
+        datasets: {},
+        annotations: {},
+    },
+    instance: null,
+    modalVisible: false,
+};
+
+export default (state: ExportState = defaultState, action: ExportActions): ExportState => {
+    switch (action.type) {
+        case ExportActionTypes.OPEN_EXPORT_MODAL:
+            return {
+                ...state,
+                modalVisible: true,
+                instance: action.payload.instance,
+            };
+        case ExportActionTypes.CLOSE_EXPORT_MODAL:
+            return {
+                ...state,
+                modalVisible: false,
+                instance: null,
+            };
+        case ExportActionTypes.EXPORT_DATASET: {
+            const { instance, format, saveImages } = action.payload;
+            let activities;
+            if (instance instanceof core.classes.Project) {
+                activities = saveImages ? state.projects.datasets : state.projects.annotations;
+            } else {
+                activities = saveImages ? state.tasks.datasets : state.tasks.annotations;
+            }
+
+            activities[instance.id] =
+                instance.id in activities && !activities[instance.id].includes(format) ?
+                    [...activities[instance.id], format] :
+                    activities[instance.id] || [format];
+
+            return {
+                ...state,
+                tasks: {
+                    datasets: instance instanceof core.classes.Task && saveImages ? activities : state.tasks.datasets,
+                    annotations:
+                        instance instanceof core.classes.Task && !saveImages ? activities : state.tasks.annotations,
+                },
+                projects: {
+                    datasets:
+                        instance instanceof core.classes.Project && saveImages ? activities : state.projects.datasets,
+                    annotations:
+                        instance instanceof core.classes.Project && !saveImages ?
+                            activities :
+                            state.projects.annotations,
```

Comment:
You didn't make a deep copy for "activities", and added property to the object, so it is already in redux storage. And on the one hand, we could just destructure all the properties like this:
tasls: {
    datasets: ...tasks.datasets,
    annotations: ...tasks.annotations,
}

But I am not sure this way is also right, because actually we should not change original storage as it done in line 47.

Reply:
Added deep copy for activities state